compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/fuse_consumer_into_loop.mlir

```diff
@@ -0,0 +1,98 @@
+    %0 = scf.forall (%arg3, %arg4) in (2, 2) shared_outs(%arg5 = %arg2) -> (tensor<64x64xf32>) {
+        //      FORALL:   %[[FIRST_LOOP:.*]] = scf.forall
+        %1 = scf.forall (%arg6, %arg7) in (2, 2) shared_outs(%arg8 = %arg5) -> (tensor<64x64xf32>) {
+            %extracted_slice = tensor.extract_slice %arg8[%arg6, %arg7] [32, 32] [1, 1] : tensor<64x64xf32> to tensor<32x32xf32>
+            %6 = linalg.matmul ins(%arg0, %arg1 : tensor<32x32xf32>, tensor<32x32xf32>) outs(%extracted_slice : tensor<32x32xf32>) -> tensor<32x32xf32>
+            scf.forall.in_parallel {
+                tensor.parallel_insert_slice %6 into %arg8[%arg6, %arg7] [32, 32] [1, 1] : tensor<32x32xf32> into tensor<64x64xf32>
+            }
+        }
+        //      FORALL:   %[[ELEM_OPERAND_2:.*]] = tensor.empty() : tensor<64x64xf32>
+        //      FORALL:   %[[ELEM_OUT:.*]] = tensor.empty() : tensor<64x64xf32>
+        //      FORALL:   %[[SECOND_LOOP:.*]]:2 = scf.forall (%[[IV0:.*]], %[[IV1:.*]]) in (2, 2) shared_outs(%[[ITER_ARG_1:.*]] = %[[FIRST_LOOP]], %[[ITER_ARG_2:.*]] = %[[ELEM_OUT]])
+        %2 = scf.forall (%arg6, %arg7) in (2, 2) shared_outs(%arg8 = %1) -> (tensor<64x64xf32>) {
+            //      FORALL:   %[[MATMUL:.*]] = linalg.matmul
+            //      FORALL:   %[[OPERAND1:.*]] = tensor.extract_slice %[[MATMUL]][%[[IV0]], %[[IV1]]] [32, 32] [1, 1]
+            //      FORALL:   %[[OPERAND2:.*]] = tensor.extract_slice %[[ELEM_OPERAND_2]][%[[IV0]], %[[IV1]]] [32, 32] [1, 1]
+            //      FORALL:   %[[OPERAND3:.*]] = tensor.extract_slice %[[ITER_ARG_2]][%[[IV0]], %[[IV1]]] [32, 32] [1, 1]
+            //      FORALL:   %[[FUSED_CONSUMER:.*]] = linalg.elemwise_binary
+            // FORALL-SAME:                      ins(%[[OPERAND1]], %[[OPERAND2]] :
+            // FORALL-SAME:                      outs(%[[OPERAND3]] :
+            //      FORALL:   scf.forall.in_parallel {
+            //      FORALL:      tensor.parallel_insert_slice %[[FUSED_CONSUMER]] into %[[ITER_ARG_2]][%[[IV0]], %[[IV1]]] [32, 32] [1, 1]
+            //      FORALL:      tensor.parallel_insert_slice %[[MATMUL]] into %[[ITER_ARG_1]][%[[IV0]], %[[IV1]]] [32, 32] [1, 1]
+            //      FORALL:   }
+            %extracted_slice = tensor.extract_slice %arg8[%arg6, %arg7] [32, 32] [1, 1] : tensor<64x64xf32> to tensor<32x32xf32>
+            %6 = linalg.matmul ins(%arg0, %arg1 : tensor<32x32xf32>, tensor<32x32xf32>) outs(%extracted_slice : tensor<32x32xf32>) -> tensor<32x32xf32>
+            scf.forall.in_parallel {
+                tensor.parallel_insert_slice %6 into %arg8[%arg6, %arg7] [32, 32] [1, 1] : tensor<32x32xf32> into tensor<64x64xf32>
+            }
+        }
+        %3 = tensor.empty() : tensor<64x64xf32>
+        %4 = tensor.empty() : tensor<64x64xf32>
+        %5 = linalg.elemwise_binary {fun = #linalg.binary_fn<add>} ins(%2, %3 : tensor<64x64xf32>, tensor<64x64xf32>) outs(%4 : tensor<64x64xf32>) -> tensor<64x64xf32>
```

Comment:
maybe I am not caught up but will this pass wotk if this was a linalg.generic doing elementwise operations in its region? If yes, can we add that as a test? If not then how do we reach this input in our codegen especially if there are two or more elementwise in a single generic op region.

Reply:
TLDR: Yes, I'll modify one of the test to use `linalg.generic` instead to demonstrate the same, sure.

More context:
Yes, it'd work on ANY consumer which implements `TilingInterface` and `DestinationPassingStyleInterace` - the caller of the upstream API only need to worry about the sending the right producer (`tensor.parallel_insert_slice` or `tensor.insert_slice`) and then the API will capture ANY consumer that satisfies the aforementioned constraint.

`linalg.generic`,`linalg.elementwise_*`, `linalg.matmul`, `linalg.softmax`, etc are few examples of the same. So, yes, it'd work.